File: Part1_CPU_CrashCourse/07_Structures/particle_simulation.f90
! 粒子模擬 - Fortran Derived Types 範例
! 使用自訂資料型態組織複雜資料

program particle_simulation
  implicit none
  
  ! 定義粒子資料型態（Derived Type）
  type :: Particle
    integer :: id                 ! 粒子編號
    real(8) :: mass              ! 質量
    real(8) :: position(3)       ! 位置 (x, y, z)
    real(8) :: velocity(3)       ! 速度 (vx, vy, vz)
    real(8) :: force(3)          ! 受力 (fx, fy, fz)
    logical :: is_active         ! 是否活躍
  end type Particle
  
  ! 變數宣告
  integer, parameter :: n_particles = 10
  type(Particle), allocatable :: particles(:)
  integer :: i
  real(8) :: dt, total_energy
  
  print *, "========================================="
  print *, "  Fortran Derived Types 範例"
  print *, "  粒子模擬系統"
  print *, "========================================="
  print *
  
  ! 配置粒子陣列
  allocate(particles(n_particles))
  
  ! 初始化粒子
  print *, "初始化", n_particles, "個粒子..."
  do i = 1, n_particles
    particles(i)%id = i
    particles(i)%mass = 1.0d0 + real(i, 8) * 0.1d0
    
    ! 隨機位置
    particles(i)%position(1) = real(i, 8) * 0.5d0
    particles(i)%position(2) = real(i, 8) * 0.3d0
    particles(i)%position(3) = real(i, 8) * 0.2d0
    
    ! 初始速度
    particles(i)%velocity(1) = sin(real(i, 8)) * 2.0d0
    particles(i)%velocity(2) = cos(real(i, 8)) * 2.0d0
    particles(i)%velocity(3) = 0.0d0
    
    ! 初始受力為零
    particles(i)%force = 0.0d0
    
    ! 所有粒子都活躍
    particles(i)%is_active = .true.
  end do
  print *, "✓ 初始化完成"
  print *
  
  ! 顯示前 5 個粒子的資訊
  print *, "前 5 個粒子的初始狀態："
  print *, "-------------------------------------------"
  do i = 1, 5
    call print_particle(particles(i))
  end do
  print *
  
  ! 簡單的時間步進模擬
  dt = 0.01d0
  print *, "執行 100 個時間步進 (dt =", dt, ")..."
  do i = 1, 100
    call update_particles(particles, n_particles, dt)
  end do
  print *, "✓ 模擬完成"
  print *
  
  ! 顯示模擬後的狀態
  print *, "粒子 #1 模擬後的狀態："
  print *, "-------------------------------------------"
  call print_particle(particles(1))
  print *
  
  ! 計算系統總能量
  total_energy = compute_total_energy(particles, n_particles)
  write(*, '(A, F12.4)') "系統總能量: ", total_energy
  print *
  
  ! 釋放記憶體
  deallocate(particles)
  
  print *, "========================================="
  print *, "  模擬完成"
  print *, "========================================="
  
contains

  ! 印出粒子資訊
  subroutine print_particle(p)
    type(Particle), intent(in) :: p
    
    write(*, '(A, I3)') "粒子 ID:    ", p%id
    write(*, '(A, F8.4)') "質量:       ", p%mass
    write(*, '(A, 3F10.4)') "位置:       ", p%position
    write(*, '(A, 3F10.4)') "速度:       ", p%velocity
    write(*, '(A, L2)') "活躍狀態:   ", p%is_active
    print *
  end subroutine print_particle
  
  ! 更新粒子狀態
  subroutine update_particles(particles, n, dt)
    type(Particle), intent(inout) :: particles(:)
    integer, intent(in) :: n
    real(8), intent(in) :: dt
    integer :: i
    
    do i = 1, n
      if (particles(i)%is_active) then
        ! 簡單的重力場
        particles(i)%force(3) = -9.8d0 * particles(i)%mass
        
        ! 更新速度 (v = v + a*dt)
        particles(i)%velocity = particles(i)%velocity + &
                                (particles(i)%force / particles(i)%mass) * dt
        
        ! 更新位置 (x = x + v*dt)
        particles(i)%position = particles(i)%position + &
                                particles(i)%velocity * dt
        
        ! 簡單的邊界條件
        if (particles(i)%position(3) < 0.0d0) then
          particles(i)%position(3) = 0.0d0
          particles(i)%velocity(3) = -particles(i)%velocity(3) * 0.8d0  ! 反彈
        end if
      end if
    end do
  end subroutine update_particles
  
  ! 計算系統總能量
  function compute_total_energy(particles, n) result(energy)
    type(Particle), intent(in) :: particles(:)
    integer, intent(in) :: n
    real(8) :: energy
    integer :: i
    real(8) :: ke, pe, v_squared
    
    energy = 0.0d0
    do i = 1, n
      if (particles(i)%is_active) then
        ! 動能 = 0.5 * m * v^2
        v_squared = dot_product(particles(i)%velocity, particles(i)%velocity)
        ke = 0.5d0 * particles(i)%mass * v_squared
        
        ! 位能 = m * g * h
        pe = particles(i)%mass * 9.8d0 * particles(i)%position(3)
        
        energy = energy + ke + pe
      end if
    end do
  end function compute_total_energy

end program particle_simulation

File: Part1_CPU_CrashCourse/07_Structures/particle_simulation.cpp
// 粒子模擬 - C++ Struct/Class 範例
// 使用自訂資料結構組織複雜資料

#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>

// 定義粒子結構 (struct)
struct Particle {
  int id;                    // 粒子編號
  double mass;               // 質量
  double position[3];        // 位置 (x, y, z)
  double velocity[3];        // 速度 (vx, vy, vz)
  double force[3];           // 受力 (fx, fy, fz)
  bool is_active;            // 是否活躍
  
  // 建構子（初始化）
  Particle(int id_val = 0, double mass_val = 1.0) 
    : id(id_val), mass(mass_val), is_active(true) {
    for (int i = 0; i < 3; i++) {
      position[i] = 0.0;
      velocity[i] = 0.0;
      force[i] = 0.0;
    }
  }
  
  // 成員函數：印出粒子資訊
  void print() const {
    std::cout << "粒子 ID:    " << id << std::endl;
    std::cout << "質量:       " << std::fixed << std::setprecision(4) 
              << mass << std::endl;
    std::cout << "位置:       " 
              << std::setw(10) << position[0] << " "
              << std::setw(10) << position[1] << " "
              << std::setw(10) << position[2] << std::endl;
    std::cout << "速度:       " 
              << std::setw(10) << velocity[0] << " "
              << std::setw(10) << velocity[1] << " "
              << std::setw(10) << velocity[2] << std::endl;
    std::cout << "活躍狀態:   " << (is_active ? "是" : "否") << std::endl;
    std::cout << std::endl;
  }
  
  // 成員函數：更新粒子狀態
  void update(double dt) {
    if (!is_active) return;
    
    // 簡單的重力場
    force[2] = -9.8 * mass;
    
    // 更新速度 (v = v + a*dt)
    for (int i = 0; i < 3; i++) {
      velocity[i] += (force[i] / mass) * dt;
    }
    
    // 更新位置 (x = x + v*dt)
    for (int i = 0; i < 3; i++) {
      position[i] += velocity[i] * dt;
    }
    
    // 簡單的邊界條件（地板反彈）
    if (position[2] < 0.0) {
      position[2] = 0.0;
      velocity[2] = -velocity[2] * 0.8;  // 反彈且損失能量
    }
  }
  
  // 成員函數：計算動能
  double kinetic_energy() const {
    double v_squared = velocity[0] * velocity[0] +
                       velocity[1] * velocity[1] +
                       velocity[2] * velocity[2];
    return 0.5 * mass * v_squared;
  }
  
  // 成員函數：計算位能
  double potential_energy() const {
    return mass * 9.8 * position[2];
  }
};

// 計算系統總能量
double compute_total_energy(const std::vector<Particle>& particles) {
  double total = 0.0;
  for (const auto& p : particles) {
    if (p.is_active) {
      total += p.kinetic_energy() + p.potential_energy();
    }
  }
  return total;
}

int main() {
  std::cout << "=========================================" << std::endl;
  std::cout << "  C++ Struct/Class 範例" << std::endl;
  std::cout << "  粒子模擬系統" << std::endl;
  std::cout << "=========================================" << std::endl;
  std::cout << std::endl;
  
  // 建立粒子陣列
  const int n_particles = 10;
  std::vector<Particle> particles;
  
  // 初始化粒子
  std::cout << "初始化 " << n_particles << " 個粒子..." << std::endl;
  for (int i = 0; i < n_particles; i++) {
    Particle p(i+1, 1.0 + i * 0.1);
    
    // 設定初始位置
    p.position[0] = i * 0.5;
    p.position[1] = i * 0.3;
    p.position[2] = i * 0.2;
    
    // 設定初始速度
    p.velocity[0] = std::sin(static_cast<double>(i)) * 2.0;
    p.velocity[1] = std::cos(static_cast<double>(i)) * 2.0;
    p.velocity[2] = 0.0;
    
    particles.push_back(p);
  }
  std::cout << "✓ 初始化完成" << std::endl;
  std::cout << std::endl;
  
  // 顯示前 5 個粒子的資訊
  std::cout << "前 5 個粒子的初始狀態：" << std::endl;
  std::cout << "-------------------------------------------" << std::endl;
  for (int i = 0; i < 5; i++) {
    particles[i].print();
  }
  
  // 簡單的時間步進模擬
  double dt = 0.01;
  std::cout << "執行 100 個時間步進 (dt = " << dt << ")..." << std::endl;
  for (int step = 0; step < 100; step++) {
    for (auto& p : particles) {
      p.update(dt);
    }
  }
  std::cout << "✓ 模擬完成" << std::endl;
  std::cout << std::endl;
  
  // 顯示模擬後的狀態
  std::cout << "粒子 #1 模擬後的狀態：" << std::endl;
  std::cout << "-------------------------------------------" << std::endl;
  particles[0].print();
  
  // 計算系統總能量
  double total_energy = compute_total_energy(particles);
  std::cout << "系統總能量: " << std::fixed << std::setprecision(4) 
            << total_energy << std::endl;
  std::cout << std::endl;
  
  std::cout << "=========================================" << std::endl;
  std::cout << "  模擬完成" << std::endl;
  std::cout << "=========================================" << std::endl;
  
  return 0;
}
